providers/duo/html.go

```diff
@@ -0,0 +1,80 @@
+package duo
+
+import (
+	"net/url"
+
+	"golang.org/x/net/html"
+)
+
+type Form struct {
+	Method string
+	Inputs map[string]string
+}
+
+func (f *Form) Set(key, value string) {
+	if f.Inputs == nil {
+		f.Inputs = make(map[string]string)
+	}
+
+	f.Inputs[key] = value
+}
+
+func (f Form) Values() url.Values {
+	v := url.Values{}
+	for key, val := range f.Inputs {
+		v.Set(key, val)
+	}
+	return v
+}
+
+func getAttribute(attrs []html.Attribute, key string) (string, bool) {
+	for _, attr := range attrs {
+		if attr.Key == key {
+			return attr.Val, true
+		}
+	}
+
+	return "", false
+}
+
+func collectFormValues(node *html.Node) (Form, error) {
+	var f Form
+	if node.Type != html.ElementNode || node.Data != "form" {
+		panic("invalid element given to parseForm")
+	}
+
+	f.Method, _ = getAttribute(node.Attr, "method")
+	for node := node.FirstChild; node != nil; node = node.NextSibling {
```

Comment:
As far as I can see, this can't find inputs if they are a bit dipper in the tree like

```
<form ...>
    <input ...> // will be found
    <div>
         <input ...> // won't be found
   </div>
</form?
```

Not sure if that's a bid deal though.

Reply:
Squashed it. Would rather fix this now than be paged because the html format or order changed a bit.